lib/reporter.js

```diff
@@ -218,58 +218,77 @@ var CoverageReporter = function (rootConfig, helper, logger, emitter) {
-          checkedCoverage[browser.id] = true
-          var coverageFailed = checkCoverage(browser, remappedCoverageMap)
-          if (coverageFailed && results) {
-            results.exitCode = 1
-          }
-        }
+    const mainDir = reporterConfig.dir || config.dir
+    const subDir = reporterConfig.subdir || config.subdir
+    const outputPath = generateOutputPath(basePath, browser.name, mainDir, subDir)
+    const remappedCoverageMap = await sourceMapStore.transformCoverage(coverageMap)
+
+    const options = helper.merge(config, reporterConfig, {
+      dir: outputPath,
+      subdir: '',
+      browser: browser,
+      emitter: emitter,
+      coverageMap: remappedCoverageMap
+    })
 
-        const context = istanbulLibReport.createContext(options)
-        const report = reports.create(reporterConfig.type || 'html', options)
+    // If config.check is defined, check coverage levels for each browser
+    if (hasOwnProperty.call(config, 'check') && !checkedCoverage[browser.id]) {
+      checkedCoverage[browser.id] = true
+      var coverageFailed = checkCoverage(browser, remappedCoverageMap)
+      if (coverageFailed && results) {
+        results.exitCode = 1
+      }
+    }
 
-        // // If reporting to console or in-memory skip directory creation
-        const toDisk = !reporterConfig.type || !reporterConfig.type.match(/^(text|text-summary|in-memory)$/)
+    const context = istanbulLibReport.createContext(options)
+    const report = reports.create(reporterConfig.type || 'html', options)
 
-        if (!toDisk && reporterConfig.file === undefined) {
-          report.execute(context)
-          return
-        }
+    // // If reporting to console or in-memory skip directory creation
+    const toDisk = !reporterConfig.type || !reporterConfig.type.match(/^(text|text-summary|in-memory)$/)
 
-        helper.mkdirIfNotExists(outputPath, function () {
-          log.debug('Writing coverage to %s', outputPath)
-          report.execute(context)
-        })
-      }))
+    if (!toDisk && reporterConfig.file === undefined) {
+      report.execute(context)
+      return
     }
+
+    helper.mkdirIfNotExists(outputPath, function () {
+      log.debug('Writing coverage to %s', outputPath)
+      report.execute(context)
+    })
+    return results
   }
 
-  this.onExit = function (done) {
+  this.onRunComplete = function (browsers) {
+    checkedCoverage = {}
+    let results = { exitCode: 0 }
+
+    const promiseCollection = reporters.map(reporterConfig =>
+      Promise.all(browsers.map(async (browser) => {
+        const res = await this.executeReport(reporterConfig, browser)
+        if (res && res.exitCode === 1) {
+          results = res
+        }
+      })))
+    promiseComplete = Promise.all(promiseCollection).then(() => results)
+    return promiseComplete
+  }
+
+  this.onExit = async function (done) {
+    const results = await promiseComplete
```

Comment:
There is no error handling by karma in case `promiseComplete` rejects. Karma doesn't expect the `onExit` handlers to return a promise. I would suggest using plain old `then` to handle success/error. See my code in #420

Reply:
await works with success only, if will be error the code onExit will be failed

Reply:
`onExit` is called sync by the karma-runner and expects the callback to be invoked. It doesn't expect a promise to be returned. That's why I suspect that a rejection would result into an unhandled Promise rejection.